### PR TITLE
Add validation for vnet and subnet cidr blocks

### LINKED ISSUE #1311: Improve validation for vnet/subnet cidr blocks
/kind bug

**What steps did you take and what happened:**
Scenario:
1. Cluster created with vnet configured with custom cidr blocks, but subnet left as default.
2. Cluster passes validation, but fails during reconciliation because the default subnet cidr does not comply with the vnet's cidr

**What did you expect to happen:**
Fail fast during cluster creation. There are two possible solutions:
1. Validate the cidr values of vnet and subnets in webhooks
2. All or nothing customization for cidr values. 

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- cluster-api-provider-azure version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by shysank**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Adds validation for vnet and subnet cidr blocks.  
- validates cidr format
- for subnet, validates if the ip is within vnet cidr 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1311 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [x] squashed commits
- [ ] includes documentation
- [x] adds unit tests

**Release note**:

```release-note
add validation for vnet and subnet cidr blocks
```

### CODE REVIEW

File: api/v1alpha4/azurecluster_validation.go [hunk 1]
```diff
@@ -182,6 +186,50 @@ func validateSubnetName(name string, fldPath *field.Path) *field.Error {
...
+	for _, vnetCidr := range vnetCidrBlocks {
+		if _, vnetNw, err := net.ParseCIDR(vnetCidr); err == nil {
+			vnetNws = append(vnetNws, vnetNw)
+		}
+	}
+
+	for _, subnetCidr := range subnetCidrBlocks {
+		subnetCidrIP, _, err := net.ParseCIDR(subnetCidr)
+		if err != nil {
+			allErrs = append(allErrs, field.Invalid(fldPath, subnetCidr, "invalid cidr format"))
```
Comment (nader-ziada) [inline api/v1alpha4/azurecluster_validation.go:203]:
```suggestion
			allErrs = append(allErrs, field.Invalid(fldPath, subnetCidr, "invalid CIDR format"))
```

File: api/v1alpha4/azurecluster_validation.go [hunk 2]
```diff
@@ -182,6 +186,50 @@ func validateSubnetName(name string, fldPath *field.Path) *field.Error {
...
+		var found bool
+		for _, vnetNw := range vnetNws {
+			if vnetNw.Contains(subnetCidrIP) {
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			allErrs = append(allErrs, field.Invalid(fldPath, subnetCidr, "subnet cidr not in vnet cidr range"))
```
Comment (nader-ziada) [inline api/v1alpha4/azurecluster_validation.go:215]:
```suggestion
			allErrs = append(allErrs, field.Invalid(fldPath, subnetCidr, "subnet CIDR not in Vnet CIDR range"))
```

File: api/v1alpha4/azurecluster_validation.go [hunk 3]
```diff
@@ -182,6 +186,50 @@ func validateSubnetName(name string, fldPath *field.Path) *field.Error {
...
+
+	return allErrs
+}
+
+// validateVnetCIDR validates the CIDR blocks of a Vnet.
+func validateVnetCIDR(vnetCIDRBlocks []string, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	for _, vnetCidr := range vnetCIDRBlocks {
+		if _, _, err := net.ParseCIDR(vnetCidr); err != nil {
+			allErrs = append(allErrs, field.Invalid(fldPath, vnetCidr, "invalid cidr format"))
```
Comment (nader-ziada) [inline api/v1alpha4/azurecluster_validation.go:227]:
```suggestion
			allErrs = append(allErrs, field.Invalid(fldPath, vnetCidr, "invalid CIDR format"))
```

**Review by nader-ziada — COMMENTED** [review]
minor comments

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm
/assign @nader-ziada

**Comment by nader-ziada** [comment]
@shysank @CecileRobertMichon is this a candidate for back-porting?

**Comment by CecileRobertMichon** [comment]
> @shysank @CecileRobertMichon is this a candidate for back-porting?

I don't think we should backport, it's technically a breaking change, even if it's a very welcome one.

**Comment by nader-ziada** [comment]
> > @shysank @CecileRobertMichon is this a candidate for back-porting?
> 
> I don't think we should backport, it's technically a breaking change, even if it's a very welcome one.

yeah good point, was just wondering

**Pull request merged**